tests/unit/sharded_copying_manager_tests/copying_manager_worker_test.py

```diff
@@ -0,0 +1,695 @@
+            copy_at_index_zero=copy_at_index_zero,
+            create_log_processor=self._instance.create_and_schedule_new_log_processor,
+        )
+
+        return processors[0]
+
+
+class TestsRunning(CopyingManagerWorkerTest):
+    @skip_on_multiprocess_workers
+    def test_wait_copying(self):
+
+        run_released = threading.Event()
+
+        _, worker = self._init_worker_instance(1, auto_start=False)
+
+        original_run = worker.run
+
+        # create mock run method that can wait for signal to start
+        def delayed_run():
+            # we need to set this event to resume worker.
+            run_released.wait()
+            return original_run()
+
+        with mock.patch.object(worker, "run", wraps=delayed_run):
+            executor = ThreadPoolExecutor()
+
+            worker.start_worker(stop_at=None)
+
+            # wait for copying start in separate thread
+
+            wait_future = executor.submit(worker.wait_for_copying_to_begin)
+
+            # check for waiting is finished, it must not be finished until event is not set.
+            assert not wait_future.done()
+            time.sleep(0.1)
+            assert not wait_future.done()
+
+            run_released.set()
+
+            # now waiting  should be finished.
+            wait_future.result(timeout=3)
+            assert wait_future.done()
+
+    def test_health_check_after_error(self):
+
+        # mock the function which is called before the worker starts copying.
+
+        def failing_init_scalyr_client(*args, **kwargs):
+            raise RuntimeError("Expected error. Do not worry.")
+
+        # mock class function with the failing one.
+        # NOTE: in order to be able to mock function in multiprocess worker, we mock a class not an instance.
+        with mock.patch.object(
+            TestableCopyingManagerThreadedWorker,
+            "_init_scalyr_client",
+            failing_init_scalyr_client,
+        ), mock.patch.object(
+            Configuration,
+            "healthy_max_time_since_last_copy_attempt",
+            new_callable=mock.PropertyMock,
+        ) as config_max_time_mock:
+
+            # reduce the 'healthy_max_time_since_last_copy_attempt' option in the config to save the time
+            config_max_time_mock.return_value = 0.01
+            # recreate and start a new shared object manager to fork a process at the moment where functions are mocked.
+            self._recreate_shared_object_manager()
+
+            # NOTE: we have to disable flow control on worker
+            # because it will just be deadlocked on the exception which we are raising.
+            _, worker = self._init_worker_instance(
+                2, auto_start=False, disable_flow_control=True
+            )
+
+            worker.start_worker()
+            worker.wait_for_copying_to_begin()
+
+            # wait until worker is crashed.
+            while worker.is_alive():
+                time.sleep(0.1)
```

Comment:
Probably we need sometime out, otherwise the test is going to just hang forever if the worker never dies.